### ISSUE TITLE
SC2025 should have an exception for non-bash Bourne shells

### ISSUE THREAD

**Issue by ssokolow** [koalaman/shellcheck#834]
#### For bugs
- Rule Id: SC2025
- My shellcheck version: 0.3.2
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh

# Default ls output to colored
alias ls='ls --color'

# Set up a nice, colorized prompt
if [ "$(id -u)" -eq 0 ]; then
  PS1='[1;31m$USER@`hostname` [1;34m$PWD #[0m '
else
  PS1='[1;32m$USER@`hostname` [1;34m$PWD $[0m '
fi

# Ensure we get colors even if a .bashrc overrules us
export color_prompt=yes
```

#### Here's what shellcheck currently says:
```
In lowendscript/colors.sh line 8:
  PS1='$USER@`hostname` $PWD # '
      ^-- SC2025: Make sure all escape sequences are enclosed in \[..\] to prevent line wrapping issues


In lowendscript/colors.sh line 10:
  PS1='$USER@`hostname` $PWD $ '
      ^-- SC2025: Make sure all escape sequences are enclosed in \[..\] to prevent line wrapping issues
```

#### Here's what I wanted or expected to see:

```
```

As with many other cases, it should recognize that my inclusion of a `#!/bin/sh` shebang means that this file is intended for Bourne shells like `dash` and `mksh` ([MirBSD Korn Shell](https://www.mirbsd.org/mksh.htm)) which don't recognize bash features like `\[` and `\]` and, instead, pass them through as literal strings.

In this case, the file is the `/etc/profile.d/colors.sh` which I deploy on resource-constrained VPSes where I use `mksh` as the lightest Debian-provided alternative to bash that still has basic tab-completion and command history. (`mksh` takes about 1/3rd the memory of bash when sourcing the same Debian default rcfiles.)

At the very least, I'd expect the incompatibility with `dash`, and `mksh` to be mentioned under Exceptions at https://github.com/koalaman/shellcheck/wiki/SC2025

(Though I can confirm that `busybox ash` has had support for `\[` and `\]` since at least the 1.13.2 release on one of my ARM devices.)

**Comment by bittorf** [comment]
interesting. so there is no portable way for setting colors for all shells the same way? dash can not work with esacpe sequences, but bash must?

**Comment by ssokolow** [comment]
Well, according to the experimentation I did while writing [this script](https://gist.github.com/ssokolow/3759156) (tested in dash, bash, pdksh, ash, zsh, a then-current busybox, and the then-outdated busybox on my OpenPandora ARM handheld), if you want to do it portably, the trick is to use `\033` with the `printf` builtin.

Since you reminded me, I decided to rewrite that file (and, as a bonus, it complicated things enough for SC2025 to not realize I was using colour escapes without `\[` and `\]`):

```sh
#!/bin/sh

# Default ls and grep output to colored and make less pass colors through
alias ls='ls --color'
alias grep='grep --color=always'
alias less='less -R'

# Be as portable as possible
# (`tput` is from ncurses-bin. `type` is a builtin in every shell I know.)
if type "tput" 1>/dev/null 2>&1; then
  # If we have ncurses-bin, support esoteric terminals
  # TODO: Figure out setaf vs. setf
  icolor() { tput setaf "$1"; tput bold; }
  creset() { tput sgr0; }
else
  # A dash/mksh-compatible way to use non-literal escape characters
  icolor() { printf '\033[1;3%dm' "$1"; }
  creset() { printf '\033[0m'; }
fi

# Set up a nice, colorized prompt
# (Use colors recognizably different from my local terminal with a mix of
#  immediate and deferred evaluation for efficiency)
if [ "$(id -u)" -eq 0 ]; then
  PS1="$(icolor 5)"'$USER@$(hostname) '"$(icolor 4)"'$PWD #'"$(creset) "
else
  PS1="$(icolor 3)"'$USER@$(hostname) '"$(icolor 4)"'$PWD \$'"$(creset) "
fi

# Ensure we get colors even if a .bashrc overrules us
export color_prompt=yes
```

**Comment by ssokolow** [comment]
What the heck. I decided to just bite the bullet and do it. While I only tested it on bash, dash, busybox sh, and mksh, this should be pretty portable:

```sh
#!/bin/sh

# Default ls and grep output to colored and make less pass colors through
alias ls='ls --color'
alias grep='grep --color=always'
alias less='less -R'

# Support bash character-printability hints
if [ -n "$BASH" ]; then
  cstart='\['
  cend='\]'
else
  cstart=''
  cend=''
fi

# Be as portable as possible
# (`tput` is from ncurses-bin. `type` is a builtin in every shell I know.)
if type "tput" 1>/dev/null 2>&1; then
  # If we have ncurses-bin, support esoteric terminals
  # TODO: Figure out setaf vs. setf
  icolor() { printf '%s%s%s' "$cstart" "$(tput setaf "$1"; tput bold)" "$cend"; }
  creset() { printf '%s%s%s' "$cstart" "$(tput sgr0)" "$cend"; }
else
  # A dash/mksh-compatible way to use non-literal escape characters
  icolor() { printf '%s\033[1;3%dm%s' "$cstart" "$1" "$cend"; }
  creset() { printf '%s\033[0m%s' "$cstart" "$cend"; }
fi

# Set up a nice, colorized prompt
# (Use colors recognizably different from my local terminal with a mix of
#  immediate and deferred evaluation for efficiency)
if [ "$(id -u)" -eq 0 ]; then
  PS1="$(icolor 5)"'$USER@$(hostname) '"$(icolor 4)"'$PWD #'"$(creset) "
else
  PS1="$(icolor 3)"'$USER@$(hostname) '"$(icolor 4)"'$PWD \$'"$(creset) "
fi

# Ensure we get colors even if a .bashrc overrules us
export color_prompt=yes
```

**Comment by kurahaupo** [comment]
> alias grep='grep --color=always'

Just for the record, `--color=always` is not such a good idea, especially if you're redefining an existing command, because it means that

    grep foo <somefile | grep -v foobar

won't behave as expected (because `foo\e[mbar` doesn't match `foobar`).

Recommendation: stick with `--color=auto`; let the user override it on a case-by-case basis.

**Comment by ssokolow** [comment]
I'm the user and, in this context, it's more important for `grep | less` to preserve coloring than for `grep | grep` to work in situations where the matches overlap.

**Comment by kurahaupo** [comment]
My comment wasn't actually directed at ssokolow; rather it's directed at anyone who might see that code in the future and decide to adopt it as their own. (Once something is published, "I'm the only user" no longer applies; it may be picked up by someone else with less understanding, maybe years in the future.)

So my further advice for posterity is to define your own custom commands, not to redefine existing ones. One way would be like this:

    c() { "$1" --color=always "${@:2}" ; }

    c grep foo < bar | less

**Comment by koalaman** [comment]
You're right. This is a bash specific tip, and should not be brought up for sh/dash/mksh/ksh.

Fixed in f0e0d9f
